### DISCUSSION TITLE
Spinner to illustrate background program is ongoing

### DISCUSSION THREAD

**Discussion by johancj** [zauberzeug/nicegui#695] [Q&A] [answered]
### Question

Hi,

I have code which uses a long time to perform (minutes). I would like the user to have a visual indication that code is running in the background with a ui.spinner. This is what I've tried:

``` python
from nicegui import ui
import time

ui.label('Hello')
with ui.dialog() as loadingDialog:
        ui.spinner(size='10em', color='black')

def wait(seconds):
    start = time.time()
    while time.time() < start + seconds:
        pass

async def someFunc():
    loadingDialog.open()
    wait(3)
    loadingDialog.close()

ui.button('Press', on_click=someFunc)

ui.run()
```

where ´´´wait()´´´ represents code that takes a lot of time to complete.

How can I make this work as intended? 

Best regards,
Johan Chr Jenssen

**Comment by pythonormen** [comment]


**Reply by johancj** [reply]
So the wait function has to be an async function in order to work? That's the only difference right? 

In my code the slow function is a sync function, but I may have to change it in order to make this work?

**Comment by falkoschindler** [comment, marked as answer]
The while loop _blocks_ until the time has passed. It even uses 100% CPU, since there is no `sleep` command:
```py
def wait(seconds):
    start = time.time()
    while time.time() < start + seconds:
        pass
```

If you want to sleep for a certain duration, you should use `asyncio.sleep` so that the UI stays responsive meanwhile.

If you have a function that does heavy computation, you should [run it in an executor](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor).

**Reply by johancj** [reply]
Thank you! It worked nicely with that link 😄

**Comment by rodja** [comment]
Just for reference: We are doing something similar in the [ffmpeg example](https://github.com/zauberzeug/nicegui/blob/main/examples/ffmpeg_extract_images/main.py).
